src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,156 @@
+            linkFirst(value);
+        } else {
+            indexValidator(index);
+            Node<T> currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            newNode = new Node<>(currentNode.previous, value, currentNode);
+            currentNode.previous.next = newNode;
+            currentNode.previous = newNode;
+        }
+        ++size;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidator(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexValidator(index);
+        Node<T> node = getNode(index);
+        T returnValue = node.value;
+        node.value = value;
+        return returnValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidator(index);
+        Node<T> nodeAtIndex = getNode(index);
+        unlinkNode(nodeAtIndex);
+        return nodeAtIndex.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if ((object == null && currentNode.value == null)
+                    || (currentNode.value != null && currentNode.value.equals(object))) {
+                unlinkNode(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> currentNode;
+        if (index < size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size; i > index + 1; i--) {
+                currentNode = currentNode.previous;
+            }
+        }
+        return currentNode;
+    }
+
+    private void linkFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        head.previous = newNode;
+        head = newNode;
+    }
+
+    private void linkLast(T value) {
+        if (size == 0) {
+            head = tail = new Node<>(null, value, null);
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+    }
+
+    private void unlinkNode(Node<T> node) {
+        if (node == tail) {
+            tail.next = null;
+            tail = node.previous;
```

Comment:
does it work correctly?
we should firstly do:
`tail = node.previous;`
and then:
`tail.next = null;`